File: frontend-web/src/utils/optimisticUI.js
/**
 * Optimistic UI pour feedback instantané
 * Met à jour l'UI immédiatement avant confirmation serveur
 */
import { useState, useCallback } from 'react';

export function useOptimisticUpdate(initialState, updateFn, rollbackFn) {
  const [state, setState] = useState(initialState);
  const [pendingUpdates, setPendingUpdates] = useState(new Map());

  const optimisticUpdate = useCallback(async (update, optimisticState) => {
    const updateId = Date.now().toString();
    
    // Mettre à jour immédiatement (optimistic)
    setState(optimisticState);
    setPendingUpdates(prev => new Map(prev).set(updateId, { update, rollback: () => setState(state) }));

    try {
      // Exécuter la mise à jour réelle
      const result = await updateFn(update);
      
      // Confirmer la mise à jour
      setState(result);
      setPendingUpdates(prev => {
        const newMap = new Map(prev);
        newMap.delete(updateId);
        return newMap;
      });

      return result;
    } catch (error) {
      // Rollback en cas d'erreur
      const pending = pendingUpdates.get(updateId);
      if (pending && rollbackFn) {
        rollbackFn(pending.update, error);
      } else {
        setState(state); // Rollback simple
      }
      
      setPendingUpdates(prev => {
        const newMap = new Map(prev);
        newMap.delete(updateId);
        return newMap;
      });

      throw error;
    }
  }, [updateFn, rollbackFn, state, pendingUpdates]);

  return [state, optimisticUpdate, pendingUpdates.size > 0];
}

/**
 * Hook pour actions optimistes simples
 */
export function useOptimisticAction(actionFn) {
  const [isPending, setIsPending] = useState(false);
  const [error, setError] = useState(null);

  const execute = useCallback(async (...args) => {
    setIsPending(true);
    setError(null);

    try {
      const result = await actionFn(...args);
      setIsPending(false);
      return result;
    } catch (err) {
      setError(err);
      setIsPending(false);
      throw err;
    }
  }, [actionFn]);

  return [execute, isPending, error];
}

File: mobile-app/lib/utils/optimistic_ui.dart
/// Optimistic UI pour feedback instantané
/// Met à jour l'UI immédiatement avant confirmation serveur
import 'package:flutter/foundation.dart';

class OptimisticUpdate<T> {
  final T optimisticValue;
  final Future<T> Function() updateFn;
  final void Function(T, dynamic)? rollbackFn;
  final ValueNotifier<T> state;
  final ValueNotifier<bool> isPending;
  final ValueNotifier<dynamic> error;

  OptimisticUpdate({
    required T initialValue,
    required this.updateFn,
    this.rollbackFn,
  })  : optimisticValue = initialValue,
        state = ValueNotifier<T>(initialValue),
        isPending = ValueNotifier<bool>(false),
        error = ValueNotifier<dynamic>(null);

  /// Exécuter une mise à jour optimiste
  Future<T?> execute(T optimisticState) async {
    // Sauvegarder l'état actuel pour rollback
    final previousState = state.value;

    // Mettre à jour immédiatement (optimistic)
    state.value = optimisticState;
    isPending.value = true;
    error.value = null;

    try {
      // Exécuter la mise à jour réelle
      final result = await updateFn();

      // Confirmer la mise à jour
      state.value = result;
      isPending.value = false;

      return result;
    } catch (e) {
      // Rollback en cas d'erreur
      if (rollbackFn != null) {
        rollbackFn!(previousState, e);
      } else {
        state.value = previousState; // Rollback simple
      }

      isPending.value = false;
      error.value = e;

      rethrow;
    }
  }

  void dispose() {
    state.dispose();
    isPending.dispose();
    error.dispose();
  }
}

/// Hook pour actions optimistes simples
class OptimisticAction {
  final ValueNotifier<bool> isPending = ValueNotifier<bool>(false);
  final ValueNotifier<dynamic> error = ValueNotifier<dynamic>(null);

  Future<T?> execute<T>(Future<T> Function() actionFn) async {
    isPending.value = true;
    error.value = null;

    try {
      final result = await actionFn();
      isPending.value = false;
      return result;
    } catch (e) {
      error.value = e;
      isPending.value = false;
      rethrow;
    }
  }

  void dispose() {
    isPending.dispose();
    error.dispose();
  }
}


